src/hotspot/cpu/ppc/abstractInterpreter_ppc.cpp

```diff
@@ -128,6 +128,15 @@ void AbstractInterpreter::layout_activation(Method* method,
     caller->interpreter_frame_esp() + caller_actual_parameters :
     caller->sp() + method->max_locals() - 1 + (frame::java_abi_size / Interpreter::stackElementSize);
 
+#ifdef ASSERT
+  if (caller->is_interpreted_frame()) {
+    assert(locals_base <= caller->interpreter_frame_expression_stack(), "bad placement");
+    // Test caller-aligned placement vs callee-aligned
+    intptr_t* l2 = caller->sp() + method->max_locals() - 1 + (frame::java_abi_size / Interpreter::stackElementSize);
+    assert(l2 >= locals_base, "bad placement");
```

Comment:
The assertion at L136 fails on ppc64 (similar to what @offamitkumar reported for s390x).
I don't understand the assertion because it is just a stricter version of the fist one.
On ppc64 the sp of `caller` is aligned down because it needs to be 16 byte aligned. `locals_base` is only 8 byte aligned. But from what I saw the difference was larger then just one word. Maybe `caller` has got an c2i extension? I guess this would be problematic.
On x86_64 `l2` depends on the last expression stack pointer not on the `caller`'s sp. If you try to translate this to ppc64 then you'll get the expression used to initialize `locals_base` at L128.
I think you can remove the 2nd assertion. Even the first one looks redundant.
Besides that I've tested `MHDeoptTest.java` successfully on ppc64.

Reply:
> Maybe `caller` has got an c2i extension? I guess this would be problematic.

I meant i2c extension.

Reply:
The two asserts together are supposed to be an upper and lower bound.  The first assert is a stricter version of the assert that was originally added by JDK-7090904.  It looks like the 2nd assert should have been reversed, assuming l2 is correct.

Reply:
I was lazy about naming, so `l2` has a different meaning in the x64 asserts.